test/vmtests/vmtests/utils/libvirt_vm.py

```diff
@@ -0,0 +1,96 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import logging
+import time
+from typing import Any, Optional
+
+import libvirt  # type: ignore
+
+
+# Assists with creating and destroying a libvirt VM.
+class LibvirtVm:
+    def __init__(self, vm_name: str, domain_xml: str, libvirt_conn: libvirt.virConnect):
+        self.vm_name: str = vm_name
+        self.domain: libvirt.virDomain = None
+
+        self.domain = libvirt_conn.defineXML(domain_xml)
+
+    def start(self) -> None:
+        # Start the VM in the paused state.
+        # This gives the console logger a chance to connect before the VM starts.
+        self.domain.createWithFlags(libvirt.VIR_DOMAIN_START_PAUSED)
+
+        # PLACEHOLDER
+        # Attach the console logger
+        # self.console_logger = LibvirtConsoleLogger()
+        # self.console_logger.attach(domain, console_log_file_path)
+
+        # Start the VM.
+        self.domain.resume()
+
+    # Wait for the VM to boot and then get the IP address.
+    def get_vm_ip_address(self, timeout: float = 30) -> str:
+        timeout_time = time.time() + timeout
+
+        while True:
+            addr = self.try_get_vm_ip_address()
+            if addr:
+                return addr
+
+            if time.time() > timeout_time:
+                raise Exception(f"No IP addresses found for '{self.vm_name}'. OS might have failed to boot.")
+
+    # Try to get the IP address of the VM.
+    def try_get_vm_ip_address(self) -> Optional[str]:
+        assert self.domain
+
+        # Acquire IP address from libvirt's DHCP server.
+        interfaces = self.domain.interfaceAddresses(libvirt.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE)
+        if len(interfaces) < 1:
+            return None
+
+        interface_name = next(iter(interfaces))
+        addrs = interfaces[interface_name]["addrs"]
+        if len(addrs) < 1:
+            return None
+
+        addr = addrs[0]["addr"]
+        assert isinstance(addr, str)
+        return addr
+
+    def close(self) -> None:
```

Comment:
nit: would it be more clear to have two separate funcitons: `stop` and `delete`? The `close` can be the destructor calling both `stop` and `delete`.

Reply:
There are actually 2 stop methods that libvirt exposes: graceful and default. So, if were to add a `stop` method, should I expose both or only one of them?

In addition, what should be the exception behavior of the `stop` and `delete` functions? Should they be best effort (like `close` is) or should they fail fast on the first error that is hit?

To be honest, I am not sure this test suite will ever need the ability to `stop` the VM outside of cleanup. So, I am inclined to keep it simple for now. If we ever need the extra functions, we can deal with the complexity then.